contracts/actions/mcd/McdBoostComposite.sol

```diff
@@ -0,0 +1,213 @@
+            _returnValues
+        );
+
+        boostParams.joinAddr = _parseParamAddr(
+            boostParams.joinAddr,
+            _paramMapping[1],
+            _subData,
+            _returnValues
+        );
+
+        boostParams.exchangeData.srcAddr = _parseParamAddr(
+            boostParams.exchangeData.srcAddr,
+            _paramMapping[2],
+            _subData,
+            _returnValues
+        );
+        boostParams.exchangeData.destAddr = _parseParamAddr(
+            boostParams.exchangeData.destAddr,
+            _paramMapping[3],
+            _subData,
+            _returnValues
+        );
+
+        boostParams.exchangeData.srcAmount = _parseParamUint(
+            boostParams.exchangeData.srcAmount,
+            _paramMapping[4],
+            _subData,
+            _returnValues
+        );
+
+        (bytes memory logData, uint256 suppliedAmount) = _boost(boostParams);
+        emit ActionEvent("McdBoostComposite", logData);
+
+        return bytes32(suppliedAmount);
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes memory _callData)
+        public
+        payable
+        virtual
+        override(ActionBase, DFSSell, GasFeeTaker)
+    {
+        BoostParams memory boostParams = _parseCompositeParams(_callData);
+        (bytes memory logData, ) = _boost(boostParams);
+        logger.logActionDirectEvent("McdBoostComposite", logData);
+    }
+
+    /// @notice Executes boost logic
+    function _boost(BoostParams memory _boostParams)
+        internal
+        returns (bytes memory logData, uint256 supplyAmount)
+    {
+        assert(_boostParams.exchangeData.srcAddr == DAI_ADDR);
+        (address urn, bytes32 ilk) = getUrnAndIlk(MCD_MANAGER_ADDR, _boostParams.vaultId);
+
+        uint256 collateralBefore;
+        if (_boostParams.gasUsed != 0) {
+            (collateralBefore, ) = getCdpInfo(
+                    IManager(MCD_MANAGER_ADDR),
+                    _boostParams.vaultId,
+                    ilk
+                );
+        }
+
+        address collateralAsset = _boostParams.exchangeData.destAddr;
+        uint256 boostAmount = _boostParams.exchangeData.srcAmount;
+
+        if (_boostParams.flAddr == address(0)) {
+            _drawDebt(_boostParams, urn, ilk);
+        }
+
+        // Sell debt asset for collateral asset
+        (uint256 exchangedAmount, ) = _dfsSell(
+            _boostParams.exchangeData,
+            address(this),
+            address(this),
+            false
+        );
+
+        // Take gas fee if part of strategy
+        if (_boostParams.gasUsed != 0) {
+            supplyAmount = _takeFee(
+                GasFeeTakerParams(
+                    _boostParams.gasUsed,
+                    collateralAsset,
+                    exchangedAmount,
+                    MAX_DFS_FEE
+                )
+            );
+        } else {
+            supplyAmount = exchangedAmount;
+        }
+
+        // Supply collateral
+        {
```

Comment:
maybe create a helper function like _drawDebt (_supplyColl) to keep code cleaner